tensorflow_lite_support/ios/task/audio/core/audio_record/sources/TFLAudioRecord.m

```diff
@@ -0,0 +1,269 @@
+                             withDomain:TFLAudioRecordErrorDomain
+                                   code:TFLAudioRecordErrorCodeProcessingError
+                            description:@"Reached end of input audio stream."];
+      break;
+    }
+    case AVAudioConverterOutputStatus_InputRanDry: {
+      [TFLCommonUtils createCustomError:error
+                             withDomain:TFLAudioRecordErrorDomain
+                                   code:TFLAudioRecordErrorCodeProcessingError
+                            description:@"Not enough input is available to satisfy the request."];
+      break;
+    }
+  }
+  return nil;
+}
+
+- (BOOL)loadAudioPCMBuffer:(AVAudioPCMBuffer *)pcmBuffer error:(NSError **)error {
+  if (pcmBuffer.frameLength == 0) {
+    [TFLCommonUtils createCustomError:error
+                           withDomain:TFLAudioRecordErrorDomain
+                                 code:TFLAudioRecordErrorCodeInvalidArgumentError
+                          description:@"You may have to try with a different "
+                                      @"channel count or sample rate"];
+  } else if ((pcmBuffer.frameLength % self.audioFormat.channelCount) != 0) {
+    [TFLCommonUtils createCustomError:error
+                           withDomain:TFLAudioRecordErrorDomain
+                                 code:TFLAudioRecordErrorCodeInvalidArgumentError
+                          description:@"You have passed an unsupported number of channels."];
+  } else {
+    TFLFloatBuffer *floatBuffer = [[TFLFloatBuffer alloc] initWithData:pcmBuffer.floatChannelData[0]
+                                                                  size:pcmBuffer.frameLength];
+
+    if ([self->_ringBuffer loadBuffer:floatBuffer offset:0 size:floatBuffer.size error:error]) {
+      return YES;
+    }
+  }
+  return NO;
+}
+
+- (void)startTappingMicrophoneWithError:(NSError **)error {
+  AVAudioNode *inputNode = [_audioEngine inputNode];
+  AVAudioFormat *format = [inputNode outputFormatForBus:0];
+
+  AVAudioFormat *recordingFormat =
+      [[AVAudioFormat alloc] initWithCommonFormat:AVAudioPCMFormatFloat32
+                                       sampleRate:self.audioFormat.sampleRate
+                                         channels:(AVAudioChannelCount)self.audioFormat.channelCount
+                                      interleaved:YES];
+
+  AVAudioConverter *audioConverter = [[AVAudioConverter alloc] initFromFormat:format
+                                                                     toFormat:recordingFormat];
+
+  // Setting buffer size takes no effect on the input node. This class uses a ring buffer internally
+  // to ensure the requested buffer size.
+  [inputNode
+      installTapOnBus:0
+           bufferSize:(AVAudioFrameCount)self.bufferSize
+               format:format
+                block:^(AVAudioPCMBuffer *buffer, AVAudioTime *when) {
+                  dispatch_async(self->_conversionQueue, ^{
```

Comment:
Move this logic to a method so that you can write unit test for it without mocking the AVAudioNode.

Reply:
Fixed